From ffc8b30866879ed9ba62bd0a86fecdbd51cd3d19 Mon Sep 17 00:00:00 2001
From: Kees Cook <keescook@chromium.org>
Date: Wed, 03 Jul 2013 22:01:14 +0000
Subject: block: do not pass disk names as format strings

Disk names may contain arbitrary strings, so they must not be
interpreted as format strings.  It seems that only md allows arbitrary
strings to be used for disk names, but this could allow for a local
memory corruption from uid 0 into ring 0.

CVE-2013-2851

Signed-off-by: Kees Cook <keescook@chromium.org>
Cc: Jens Axboe <axboe@kernel.dk>
Cc: <stable@vger.kernel.org>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>

[jmm: Backport to 2.6.32]

diff -aur source.orig/drivers/block/nbd.c source/drivers/block/nbd.c
--- source.orig/drivers/block/nbd.c	2013-09-03 11:22:07.898409796 +0200
+++ source/drivers/block/nbd.c	2013-09-03 11:53:14.178696265 +0200
@@ -655,7 +655,9 @@
 
 		mutex_unlock(&lo->tx_lock);
 
-		thread = kthread_create(nbd_thread, lo, lo->disk->disk_name);
+		thread = kthread_create(nbd_thread, lo, "%s",
+					lo->disk->disk_name);
+
 		if (IS_ERR(thread)) {
 			mutex_lock(&lo->tx_lock);
 			return PTR_ERR(thread);

diff -aur source.orig/fs/partitions/check.c source/fs/partitions/check.c
--- source.orig/fs/partitions/check.c	2013-09-03 11:21:57.102292441 +0200
+++ source/fs/partitions/check.c	2013-09-03 11:51:59.833888139 +0200
@@ -476,7 +476,7 @@
 
 	ddev->parent = disk->driverfs_dev;
 
-	dev_set_name(ddev, disk->disk_name);
+	dev_set_name(ddev, "%s", disk->disk_name);
 
 	/* delay uevents, until we scanned partition table */
 	dev_set_uevent_suppress(ddev, 1);

